### PR TITLE
iox-605 Introduce `memory` namespace

### LINKED ISSUE #605: Refactor Relocatable Pointer and Relative Pointer
## Brief feature description

The relocatable pointer (`iceoryx_utils/internal/relocatable_pointer`) implementation needs to be cleaned up:

 1. [x] add noexcept, const, replace `size_t` with `uint64_t` (done in #618 except making parameters const)
 2. [x] separate implementation and declaration for all classes (#618)
 3. [x] one class per header file (#618)
 4. [x] add subnamespace `iox::rp` where everything is stored (#618)
 5. [x] ~~restructure the unit tests / add missing tests~~ Moved to #1745
 6. [x] add doxygen documentation (#618)
 7. [x] remove includes of base_relative_ptr.hpp & base_relocatable_ptr.hpp
 9. [x] ~~consider removing ``ptr_t`` and ``const_ptr_t`` aliases and use ``void*`` directly (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597604809)~~
 10. [x] ~~add ``void* get() noexcept``, ``const void* get() const noexcept``, ``void* getBasePtr() noexcept`` and ``const void* getBasePtr() const noexcept`` to BaseRelativePointer~~ -> Not implemented as `SharedPointer` uses `get() const noexcept`
 11. [x] ~~consider changing return type of ``PointerRepository::unregisterPrt()`` to ``void`` (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597609248)~~ - Note: maybe not, it might be useful to have an indication whether the deregistration was successful
 12. [x] move duplicated code into a separate method (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597613171)
 13. [x] rename ``PointerRepository::registerPtr()`` methods if both are needed (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597613855) + consider returning a ``cxx::unique_ptr<id_t>``
 14. [x] check invalid id logic (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597617681 and https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597619467)
 15. [x] ~~add const methods to RelativePointer (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597620985, https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597621144 and https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597621263).~~ This might not be needed, compare with e.g. std::shared_pointer https://en.cppreference.com/w/cpp/memory/shared_ptr/operator* It should work similar to STL smart pointers.
 16. [x] invalidate ``other`` move c'tor and move assignment operator of BaseRelocatablePointer (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597626715 and https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r597628041)
 17. [x] use composition instead of inheritance (see https://github.com/eclipse-iceoryx/iceoryx/pull/618#discussion_r599596822 and https://github.com/eclipse-iceoryx/iceoryx/issues/605#issuecomment-806045008) - if done, consider renaming Base...Pointer to Untyped...Pointer
 18. [x] ~~replace `RelativePointer` by `RelocatablePointer` where applicable, e.g. `Semaphore` and `LoFFLi`~~ Moved to #1745
 19. [x] Move `RelativePointer` to public include

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

* Rename directory from `reloctable_pointer` to `memory`
* Rename namespace `rp` to `memory`

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #605 (partly)

### CODE REVIEW

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 1]
```diff
@@ -403,19 +403,19 @@
...
     ```
 
-20. Renamed `BaseRelativePointer` to `UntypedRelativePointer`
+21. Renamed `BaseRelativePointer` to `UntypedRelativePointer`
 
     ```cpp
     // before
-    #include "iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp"
-    iox::rp::BaseRelativePointer myUntypedRelativePointer;
+    #include "iceoryx_hoofs/internal/memory/base_relative_pointer.hpp"
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/release-notes/iceoryx-unreleased.md:410]: Shouldn't the namespace renaming be a separate entry in the changelog?

File: iceoryx_hoofs/test/moduletests/test_relative_pointer.cpp [hunk 2]
```diff
@@ -78,11 +78,11 @@ TYPED_TEST(RelativePointer_test, ConstrTests)
     ::testing::Test::RecordProperty("TEST_ID", "cae7b4d4-86eb-42f6-b938-90a76f01bea5");
     // NOLINTJUSTIFICATION Pointer arithmetic needed for tests
     // NOLINTBEGIN(cppcoreguidelines-pro-bounds-pointer-arithmetic, cppcoreguidelines-pro-type-reinterpret-cast)
-    EXPECT_EQ(RelativePointer<TypeParam>::registerPtrWithId(iox::rp::segment_id_t{1U},
+    EXPECT_EQ(RelativePointer<TypeParam>::registerPtrWithId(iox::memory::segment_id_t{1U},
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_relative_pointer.cpp:81]: @FerdinandSpitzschnueffler Good catch, removed them.

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 3]
```diff
@@ -403,19 +403,19 @@
     }
     ```
 
-20. Renamed `BaseRelativePointer` to `UntypedRelativePointer`
+21. Renamed `BaseRelativePointer` to `UntypedRelativePointer` and moved it from namespace `rp::` to `memory::`
 
     ```cpp
     // before
-    #include "iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp"
+    #include "iceoryx_hoofs/internal/memory/base_relative_pointer.hpp"
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/release-notes/iceoryx-unreleased.md:410]:
```suggestion
    #include "iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp"
```

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by MatthiasKillat — APPROVED** [review]

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 4]
```diff
@@ -399,7 +400,19 @@
...
-20. The `CMakeLists.txt` of apps using iceoryx need to add `iceoryx_platform`
+21. Renamed `BaseRelativePointer` to `UntypedRelativePointer`
+
+    ```cpp
+    // before
+    #include "iceoryx_hoofs/internal/memory/base_relative_pointer.hpp"
+    iox::memory::BaseRelativePointer myUntypedRelativePointer;
+
+    // after
+    #include "iceoryx_hoofs/memory/relative_pointer.hpp"
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply doc/website/release-notes/iceoryx-unreleased.md:414]: I added a todo in #605

**Pull request merged**